tests/e2e-tests/src/recursor/dnssec/scenarios.rs

```diff
@@ -0,0 +1,242 @@
+use std::net::Ipv4Addr;
+
+use dns_test::{
+    FQDN, Implementation, Network, Resolver, Result, TrustAnchor,
+    client::{Client, DigOutput, DigSettings},
+    name_server::{Graph, NameServer, Sign},
+    record::{Record, RecordType},
+    zone_file::{Nsec, SignSettings},
+};
+
+/// This tests that the server will return an insecure answer when receiving NSEC3 records with
+/// an iteration count exceeding the default soft limit, but below the hard limit.
+#[test]
+fn soft_nsec3_iteration_failure() -> Result<()> {
+    let (output, logs) = insecure_record_fixture(
+        FQDN("noexist.insecure.testing.")?,
+        SignSettings::default().nsec(Nsec::_3 {
+            iterations: 101,
+            salt: None,
+            opt_out: false,
+        }),
+        None,
+    )?;
+
+    dbg!(&output);
+
+    assert!(output.status.is_nxdomain());
+    assert!(!output.flags.authenticated_data);
+    assert!(logs.contains("iteration count 101 is over 100"));
+
+    Ok(())
+}
+
+/// This tests that the server will return SERVFAIL when receiving NSEC3 records with
+/// an iteration count exceeding the default hard limit.
+#[test]
+fn hard_nsec3_iteration_failure() -> Result<()> {
+    let (output, logs) = insecure_record_fixture(
+        FQDN("noexist.insecure.testing.")?,
+        SignSettings::default().nsec(Nsec::_3 {
+            iterations: 501,
+            salt: None,
+            opt_out: false,
+        }),
+        None,
+    )?;
+
+    dbg!(&output);
+
+    assert!(output.status.is_servfail());
+    assert!(logs.contains("iteration count 501 is over 500"));
+
+    Ok(())
+}
+
+/// This tests the nsec3_soft_iteration_limit and nsec3_hard_iteration_limit configuration settings
+#[test]
+fn nsec3_custom_iteration_count() -> Result<()> {
+    let (output, logs) = insecure_record_fixture(
+        FQDN("noexist.insecure.testing.")?,
+        SignSettings::default().nsec(Nsec::_3 {
+            iterations: 11,
+            salt: None,
+            opt_out: false,
+        }),
+        Some(minijinja::render!(include_str!(
+            "custom_nsec3_iterations.toml.jinja"
+        ))),
+    )?;
+
+    dbg!(&output);
+
+    assert!(output.status.is_nxdomain());
+    assert!(!output.flags.authenticated_data);
+    assert!(logs.contains("iteration count 11 is over 10"));
+
+    let (output, logs) = insecure_record_fixture(
+        FQDN("noexist.insecure.testing.")?,
+        SignSettings::default().nsec(Nsec::_3 {
+            iterations: 21,
+            salt: None,
+            opt_out: false,
+        }),
+        Some(minijinja::render!(include_str!(
+            "custom_nsec3_iterations.toml.jinja"
+        ))),
+    )?;
+
+    dbg!(&output);
+
+    assert!(output.status.is_servfail());
+    assert!(logs.contains("iteration count 21 is over 20"));
+
+    Ok(())
+}
+
+/// This test verifies the server will verify NSEC3 RRSIGs before rejecting high iteration counts.
+#[test]
+fn hard_nsec3_iteration_invalid_rrsig() -> Result<()> {
```

Comment:
It would be good to have another test that combines an invalid RRSIG with an iteration count just over the soft limit. This test already checks for log lines, but if we only exceeded the soft limit, then the response code would give us a bit more assurance that the logic is doing what we want.